### PR TITLE
[IMP] autofill: support date autofill

### PR DESCRIPTION
**Pull request by rrahir**
Task: 4173602

## Description:

description of this task, what is implemented and why it is implemented that way.

Task: [TASK_ID](https://www.odoo.com/web#id=TASK_ID&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_t("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: src/registries/autofill_rules.ts [hunk 1]
```diff
@@ -77,6 +86,53 @@ function calculateIncrementBasedOnGroup(group: number[]) {
   return increment;
 }
 
+function getEqualInterval(intervals: DateIncrement[], key: keyof DateIncrement) {
```
Comment (hokolomopo) [inline src/registries/autofill_rules.ts:89]: I don't have good ideas for the name, but it was kinda hard to understand what the function does exactly. Maybe inline it/add some jsDoc ?

File: src/registries/autofill_rules.ts [hunk 2]
```diff
@@ -77,6 +86,53 @@ function calculateIncrementBasedOnGroup(group: number[]) {
...
+  const jsDates = group.map((date) => toJsDate(date, DEFAULT_LOCALE));
+
+  const datesIntervals = getDateIntervals(jsDates);
+
+  const daysInterval = getEqualInterval(datesIntervals, "days");
+  const monthsInterval = getEqualInterval(datesIntervals, "months");
+  const yearsInterval = getEqualInterval(datesIntervals, "years");
+
+  // ## algo un peu pété il faut quand meme fixer les deltas
+  // et véfier que c'est les memes tout simplement ?
```
Comment (hokolomopo) [inline src/registries/autofill_rules.ts:118]: 🥖

Comment (hokolomopo) [reply src/registries/autofill_rules.ts:118]: The algo seems _fine_, it doesn't do the same thing as Gsheet/Excel if we have an increment of 1 day and 1 year for example, but I don't really understand what they do 🤷

File: src/registries/autofill_rules.ts [hunk 3]
```diff
@@ -154,3 +253,97 @@ autofillRulesRegistry
...
+    return {
+      years,
+      months,
+      days,
+    };
+  });
+  return res.slice(1);
+}
+
+function daysBetween(date1: DateTime, date2: DateTime) {
```
Comment (hokolomopo) [inline src/registries/autofill_rules.ts:282]: We already have helpers for the difference in days/month/years between dates (DATEDIF function)

File: tests/autofill/autofill_plugin.test.ts [hunk 4]
```diff
@@ -218,14 +218,119 @@ describe("Autofill", () => {
...
+        setCellContent(model, "A3", "3/24/2006");
+        autofill("A1:A3", "A6");
+        expect(getCellText(model, "A4")).toBe("3/24/2009");
+        expect(getCellText(model, "A5")).toBe("3/24/2012");
+        expect(getCellText(model, "A6")).toBe("3/24/2015");
+      });
+
+      test("dates 2 year apart with leap year", () => {
+        setCellContent(model, "A1", "3/24/2000");
+        setCellContent(model, "A2", "3/24/2002");
```
Comment (hokolomopo) [inline tests/autofill/autofill_plugin.test.ts:284]: You should pick a date before march to test the leap year no ? There's no 29 february between the two dates you picked :)

File: src/registries/autofill_rules.ts [hunk 5]
```diff
@@ -77,6 +86,53 @@ function calculateIncrementBasedOnGroup(group: number[]) {
...
+  // ## algo un peu pété il faut quand meme fixer les deltas
+  // et véfier que c'est les memes tout simplement ?
+  const dateIntervalParts = [yearsInterval, monthsInterval, daysInterval];
+
+  const index = dateIntervalParts.findIndex((el) => !!el);
+  const lastIndex = dateIntervalParts.findIndex((el) => !!el);
+
+  const isSingleDay = dateIntervalParts.every((el) => el === 0);
+  if (index !== lastIndex || index === -1 || isSingleDay) {
+    return group.length === 2 ? group.length * (group[1] - group[0]) : undefined; // *2 because only 2 values
```
Comment (hokolomopo) [inline src/registries/autofill_rules.ts:126]: There is no `*2` in the expression 😅

**Review by hokolomopo — COMMENTED** [review]
Improving the autofill, this sounds easy 🙃

File: tests/autofill/autofill_plugin.test.ts (hunk 4 above)
Comment (rrahir) [reply tests/autofill/autofill_plugin.test.ts:284]: doesn't change anything, i'm bound to "pass though" the 29 of february either between 2002 and 2004 or 2004 and 2006

File: src/registries/autofill_rules.ts (hunk 3 above)
Comment (rrahir) [reply src/registries/autofill_rules.ts]: I figured but  failed to find them >< thanks :)

File: src/helpers/dates.ts [hunk 6]
```diff
@@ -136,6 +136,10 @@ export class DateTime {
   setSeconds(seconds: number) {
     return this.jsDate.setUTCSeconds(seconds);
   }
+
+  copy() {
+    return DateTime.fromTimestamp(this.getTime());
+  }
```
Comment (VincentSchippefilt) [inline src/helpers/dates.ts:142]: I wonder if this is a good method. it looks like a new constructor, and has only 1 usage...
Also, it only copies the time.

**Review by VincentSchippefilt — APPROVED** [review]

**Comment by rrahir** [comment]
let's go for 18.0 :o

**Review by LucasLefevre — APPROVED** [review]
> let's go for 18.0 :o

on est des fous nous :upside_down_face: 
robodoo r+

**Pull request closed without merging**